deepspeed/runtime/swap_tensor/partitioned_param_swapper.py

```diff
@@ -294,21 +305,43 @@ def swap_in(self, params, async_op=True, swap_in_buffers=None):
 
             assert len(swap_in_paths) <= len(self.available_buffer_ids), f"Not enough buffers {len(self.available_buffer_ids)} for swapping {len(swap_in_paths)}"
             compute_buffers, swap_in_buffers = self._allocate_and_return_buffers_for_swap_in(params)
+            inflight_numel = sum([t.numel() for t in compute_buffers])
+        else:
+            inflight_numel = sum([t.numel() for t in swap_in_buffers])
 
         swap_in_tensors(self.aio_read_handle, swap_in_buffers, swap_in_paths)
 
-        self.inflight_params.extend(params)
-        self.inflight_swap_in_buffers.extend(swap_in_buffers)
-        self.inflight_numel += sum([t.numel() for t in compute_buffers])
-
-        for param in params:
-            param.ds_tensor.status = PartitionedParamStatus.INFLIGHT
-
-        self.pending_reads += len(params)
+        self._update_inflight_swap_in(params, swap_in_buffers, inflight_numel)
 
         if not async_op:
             self.synchronize_reads()
 
+    # Enables swapping into buffer that is out the control of swapper. This is always synchronous
+    def swap_into_buffer(self, param, dest_buffer):
+        assert param.ds_tensor.status == PartitionedParamStatus.NOT_AVAILABLE, f"param {param.ds_id} is already available or inflight"
+
+        require_swap_buffer = not (dest_buffer.is_pinned()
+                                   and self._is_io_aligned(dest_buffer.numel()))
+
+        if require_swap_buffer:
+            assert len(self.available_buffer_ids) > 0, f"No buffer available to swap param {param.ds_id}."
+            compute_buffers, swap_in_buffers = self._allocate_and_return_buffers_for_swap_in([param])
+            inflight_numel = compute_buffers[0].numel()
+        else:
+            swap_in_buffers = [dest_buffer]
+            inflight_numel = dest_buffer.numel()
+
+        swap_in_paths = self._get_swap_paths([param])
+
+        swap_in_tensors(self.aio_read_handle, swap_in_buffers, swap_in_paths)
+        self._update_inflight_swap_in([param], swap_in_buffers, inflight_numel)
+        self.synchronize_reads()
+
+        if require_swap_buffer:
+            dest_buffer.data.copy_(param.ds_tensor.data)
```

Comment:
if dest_buffer is not pinned in memory, should we pin it here or doesn't matter?

Reply:
Good question. Since we don't own `dest_buffer`, we leave it to the caller to decide if they want to pin or not.